File: src/ui/web/src/lib/queryFilter.ts
/**
 * Filter algebra types and URL conversion for the unified query system.
 *
 * Each filter is a `LeafFilter { type, value }` matching the server's
 * `?f=prefix:value` format. The `type` field matches the server's filter
 * `prefix()` (e.g., "camera_make", "media", "iso", "favorite").
 *
 * FilterCapability describes a filter type from the server's
 * GET /v1/filters/capabilities endpoint.
 */

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** A single filter predicate. Matches server's LeafFilter serialization. */
export interface LeafFilter {
  type: string;
  value: string;
}

/** A filter type descriptor from GET /v1/filters/capabilities. */
export interface FilterCapability {
  prefix: string;
  label: string;
  value_kind: string; // "text" | "string" | "boolean" | "enum" | "int_range" | "float_range" | "date_range" | "location"
  faceted?: boolean;
  enum_values?: string[];
}

/** Saved query format for smart collections. */
export interface SavedQueryV2 {
  filters: LeafFilter[];
  sort?: string;
  direction?: string;
}

// ---------------------------------------------------------------------------
// URL ↔ Filter conversion
// ---------------------------------------------------------------------------

/**
 * Convert filters to URL search params (repeated `f=prefix:value`).
 * Sort/direction are separate top-level params.
 */
export function filtersToParams(
  filters: LeafFilter[],
  sort?: string,
  direction?: string,
): URLSearchParams {
  const params = new URLSearchParams();
  for (const f of filters) {
    params.append("f", `${f.type}:${f.value}`);
  }
  if (sort && sort !== "taken_at") params.set("sort", sort);
  if (direction && direction !== "desc") params.set("dir", direction);
  return params;
}

/**
 * Parse URL search params into filters + sort/direction.
 * Reads repeated `f=prefix:value` params.
 */
export function paramsToFilters(params: URLSearchParams): {
  filters: LeafFilter[];
  sort: string;
  direction: "asc" | "desc";
} {
  const filters: LeafFilter[] = [];
  for (const raw of params.getAll("f")) {
    const colon = raw.indexOf(":");
    if (colon <= 0) continue;
    filters.push({
      type: raw.slice(0, colon),
      value: raw.slice(colon + 1),
    });
  }
  const sort = params.get("sort") ?? "taken_at";
  const dir = params.get("dir");
  const direction: "asc" | "desc" = dir === "asc" ? "asc" : "desc";
  return { filters, sort, direction };
}

// ---------------------------------------------------------------------------
// Filter helpers
// ---------------------------------------------------------------------------

/** Get a filter's value by type, or undefined if not present. */
export function getFilterValue(filters: LeafFilter[], type: string): string | undefined {
  return filters.find((f) => f.type === type)?.value;
}

/** Check whether any user-visible filters are active (excludes sort/dir). */
export function hasActiveFilters(filters: LeafFilter[]): boolean {
  return filters.length > 0;
}

/** Set or remove a filter by type. Returns new array. */
export function setFilter(
  filters: LeafFilter[],
  type: string,
  value: string | null,
): LeafFilter[] {
  const without = filters.filter((f) => f.type !== type);
  if (value == null || value === "") return without;
  return [...without, { type, value }];
}

/** Remove all filters. */
export function clearFilters(): LeafFilter[] {
  return [];
}

// ---------------------------------------------------------------------------
// Labels
// ---------------------------------------------------------------------------

/** Human-readable label for a filter, using capabilities if available. */
export function filterLabel(
  filter: LeafFilter,
  capabilities?: FilterCapability[],
): string {
  const cap = capabilities?.find((c) => c.prefix === filter.type);
  const typeLabel = cap?.label ?? FALLBACK_LABELS[filter.type] ?? filter.type;

  switch (filter.type) {
    case "query":
      // No wrapping quotes — users can type literal `"phrase"` for
      // exact-match and the decorative outer quotes would collide
      // with the input, rendering as `""phrase""`.
      return `Search: ${filter.value}`;
    case "media":
      return filter.value === "image" ? "Photos" : filter.value === "video" ? "Videos" : `Media: ${filter.value}`;
    case "favorite":
      return filter.value === "yes" ? "Favorites" : "Not favorites";
    case "has_gps":
      return filter.value === "yes" ? "Has GPS" : "No GPS";
    case "has_faces":
      return filter.value === "yes" ? "Has faces" : "No faces";
    case "has_exposure":
      return filter.value === "yes" ? "Has exposure data" : "No exposure data";
    case "has_rating":
      return filter.value === "yes" ? "Has rating" : "No rating";
    case "has_color":
      return filter.value === "yes" ? "Has color label" : "No color label";
    case "stars": {
      if (filter.value.includes("-")) {
        const [lo, hi] = filter.value.split("-");
        return lo === hi ? `${lo} star${lo === "1" ? "" : "s"}` : `${lo}–${hi} stars`;
      }
      if (filter.value.endsWith("+")) return `${filter.value.slice(0, -1)}+ stars`;
      return `${filter.value} star${filter.value === "1" ? "" : "s"}`;
    }
    case "iso": {
      if (filter.value.includes("-")) {
        const [lo, hi] = filter.value.split("-");
        return `ISO ${lo}–${hi}`;
      }
      if (filter.value.endsWith("+")) return `ISO ${filter.value.slice(0, -1)}+`;
      return `ISO ${filter.value}`;
    }
    case "date":
      return `Date: ${filter.value.replace(",", " – ")}`;
    case "color":
      return `Color: ${filter.value}`;
    case "tag":
      return `Tag: ${filter.value}`;
    case "near":
      return "Near location";
    case "person":
      return "Person filter";
    case "library":
      return typeLabel;
    default:
      return `${typeLabel}: ${filter.value}`;
  }
}

const FALLBACK_LABELS: Record<string, string> = {
  query: "Search",
  library: "Library",
  path: "Path",
  media: "Media Type",
  camera_make: "Camera Make",
  camera_model: "Camera Model",
  lens: "Lens",
  iso: "ISO",
  aperture: "Aperture",
  focal_length: "Focal Length",
  exposure: "Exposure",
  has_exposure: "Has Exposure",
  has_gps: "Has GPS",
  near: "Near",
  date: "Date",
  favorite: "Favorite",
  stars: "Stars",
  color: "Color",
  has_rating: "Has Rating",
  has_color: "Has Color",
  has_faces: "Has Faces",
  person: "Person",
  tag: "Tag",
};

// ---------------------------------------------------------------------------
// Range / compound value parsing helpers (for FilterBar widgets)
// ---------------------------------------------------------------------------

/** Parse "200-800" or "400+" or "-1600" into {min, max}. */
export function parseRange(value: string | undefined): { min: string | null; max: string | null } {
  if (!value) return { min: null, max: null };
  if (value.endsWith("+")) return { min: value.slice(0, -1), max: null };
  if (value.startsWith("-")) return { min: null, max: value.slice(1) };
  const dash = value.indexOf("-");
  if (dash > 0) return { min: value.slice(0, dash), max: value.slice(dash + 1) };
  return { min: value, max: value }; // exact value
}

/** Compose range back to filter value. Returns null if both are empty. */
export function composeRange(min: string | null, max: string | null): string | null {
  if (!min && !max) return null;
  if (min && !max) return `${min}+`;
  if (!min && max) return `-${max}`;
  if (min === max) return min!;
  return `${min}-${max}`;
}

/** Parse "48.85,2.35,5" into {lat, lon, radius}. */
export function parseNear(value: string | undefined): { lat: string; lon: string; radius: string } | null {
  if (!value) return null;
  const parts = value.split(",");
  if (parts.length < 3) return null;
  return { lat: parts[0], lon: parts[1], radius: parts[2] };
}

/** Compose near location back to filter value. */
export function composeNear(lat: string, lon: string, radius: string): string {
  return `${lat},${lon},${radius}`;
}

/** Parse "2024-01-01,2024-12-31" into {from, to}. */
export function parseDate(value: string | undefined): { from: string | null; to: string | null } {
  if (!value) return { from: null, to: null };
  const parts = value.split(",", 2);
  return { from: parts[0] || null, to: parts[1] || null };
}

/** Compose date range back to filter value. Returns null if both are empty. */
export function composeDate(from: string | null, to: string | null): string | null {
  if (!from && !to) return null;
  return `${from ?? ""},${to ?? ""}`;
}

// ---------------------------------------------------------------------------
// Smart collection serialization
// ---------------------------------------------------------------------------

/** Build a SavedQueryV2 from the current filter state. */
export function buildSavedQuery(
  filters: LeafFilter[],
  sort: string,
  direction: string,
): SavedQueryV2 {
  return {
    filters,
    sort,
    direction,
  };
}

/** Generate human-readable labels for all filters in a saved query. */
export function savedQueryLabels(
  sq: SavedQueryV2,
  capabilities?: FilterCapability[],
): string[] {
  return sq.filters
    .filter((f) => f.type !== "library")
    .map((f) => filterLabel(f, capabilities));
}

File: clients/lumiverb-app/Sources/LumiverbKit/Sources/LumiverbKit/Models/QueryFilter.swift
/// Filter algebra types for the unified query system.
///
/// Each filter is a `LeafFilter { type, value }` matching the server's
/// `?f=prefix:value` format. The `type` field matches the server's filter
/// prefix (e.g., "camera_make", "media", "iso", "favorite").

import Foundation

// MARK: - Core Types

/// A single filter predicate. Matches the server's LeafFilter serialization.
public struct LeafFilter: Codable, Equatable, Sendable, Hashable {
    public let type: String
    public let value: String

    public init(type: String, value: String) {
        self.type = type
        self.value = value
    }
}

/// A filter type descriptor from GET /v1/filters/capabilities.
public struct FilterCapability: Codable, Equatable, Sendable {
    public let prefix: String
    public let label: String
    public let valueKind: String
    public let faceted: Bool?
    public let enumValues: [String]?

    enum CodingKeys: String, CodingKey {
        case prefix, label
        case valueKind = "value_kind"
        case faceted
        case enumValues = "enum_values"
    }
}

/// Saved query format for smart collections (new filter algebra).
public struct SavedQueryV2: Codable, Equatable, Sendable {
    public let filters: [LeafFilter]
    public let sort: String?
    public let direction: String?

    public init(filters: [LeafFilter], sort: String? = nil, direction: String? = nil) {
        self.filters = filters
        self.sort = sort
        self.direction = direction
    }
}

// MARK: - URL Conversion

/// Convert filters to URL query items (repeated `f=prefix:value`).
public func filtersToQueryItems(
    _ filters: [LeafFilter],
    sort: String? = nil,
    direction: String? = nil,
    after: String? = nil,
    limit: Int? = nil
) -> [URLQueryItem] {
    var items: [URLQueryItem] = []
    for f in filters {
        items.append(URLQueryItem(name: "f", value: "\(f.type):\(f.value)"))
    }
    if let sort, sort != "taken_at" {
        items.append(URLQueryItem(name: "sort", value: sort))
    }
    if let direction, direction != "desc" {
        items.append(URLQueryItem(name: "dir", value: direction))
    }
    if let after {
        items.append(URLQueryItem(name: "after", value: after))
    }
    if let limit {
        items.append(URLQueryItem(name: "limit", value: String(limit)))
    }
    return items
}

// MARK: - Filter Helpers

/// Get a filter's value by type, or nil if not present.
public func getFilterValue(_ filters: [LeafFilter], type: String) -> String? {
    filters.first(where: { $0.type == type })?.value
}

/// Set or remove a filter by type. Returns a new array.
public func setFilter(_ filters: [LeafFilter], type: String, value: String?) -> [LeafFilter] {
    var result = filters.filter { $0.type != type }
    if let value, !value.isEmpty {
        result.append(LeafFilter(type: type, value: value))
    }
    return result
}

/// Remove all filters.
public func clearFilters() -> [LeafFilter] {
    []
}

// MARK: - Range Helpers

/// Parse "200-800" or "400+" or "-1600" into (min, max).
public func parseRange(_ value: String?) -> (min: String?, max: String?) {
    guard let value, !value.isEmpty else { return (nil, nil) }
    if value.hasSuffix("+") {
        return (String(value.dropLast()), nil)
    }
    if value.hasPrefix("-") {
        return (nil, String(value.dropFirst()))
    }
    if let dash = value.firstIndex(of: "-"), dash != value.startIndex {
        return (String(value[..<dash]), String(value[value.index(after: dash)...]))
    }
    return (value, value)
}

/// Compose range back to filter value. Returns nil if both are nil.
public func composeRange(min: String?, max: String?) -> String? {
    switch (min, max) {
    case (nil, nil): return nil
    case (let lo?, nil): return "\(lo)+"
    case (nil, let hi?): return "-\(hi)"
    case (let lo?, let hi?) where lo == hi: return lo
    case (let lo?, let hi?): return "\(lo)-\(hi)"
    }
}

/// Compose near location to filter value.
public func composeNear(lat: String, lon: String, radius: String) -> String {
    "\(lat),\(lon),\(radius)"
}

/// Compose date range to filter value. Returns nil if both are nil.
public func composeDate(from: String?, to: String?) -> String? {
    guard from != nil || to != nil else { return nil }
    return "\(from ?? ""),\(to ?? "")"
}

// MARK: - Labels

/// Human-readable label for a filter.
public func filterLabel(_ filter: LeafFilter) -> String {
    switch filter.type {
    case "query":
        // No wrapping quotes — users can type literal `"phrase"` for
        // exact-match and the decorative outer quotes would collide
        // with the input, rendering as `""phrase""`.
        return "Search: \(filter.value)"
    case "media":
        return filter.value == "image" ? "Photos" : filter.value == "video" ? "Videos" : "Media: \(filter.value)"
    case "favorite":
        return filter.value == "yes" ? "Favorites" : "Not favorites"
    case "has_gps":
        return filter.value == "yes" ? "Has GPS" : "No GPS"
    case "has_faces":
        return filter.value == "yes" ? "Has faces" : "No faces"
    case "has_exposure":
        return filter.value == "yes" ? "Has exposure data" : "No exposure data"
    case "has_rating":
        return filter.value == "yes" ? "Has rating" : "No rating"
    case "has_color":
        return filter.value == "yes" ? "Has color label" : "No color label"
    case "stars":
        if filter.value.contains("-") {
            let parts = filter.value.split(separator: "-")
            if parts.count == 2 {
                return parts[0] == parts[1] ? "\(parts[0]) star\(parts[0] == "1" ? "" : "s")" : "\(parts[0])–\(parts[1]) stars"
            }
        }
        if filter.value.hasSuffix("+") { return "\(filter.value.dropLast())+ stars" }
        return "\(filter.value) star\(filter.value == "1" ? "" : "s")"
    case "iso":
        if filter.value.contains("-") {
            let parts = filter.value.split(separator: "-")
            return "ISO \(parts[0])–\(parts.count > 1 ? String(parts[1]) : "")"
        }
        if filter.value.hasSuffix("+") { return "ISO \(filter.value.dropLast())+" }
        return "ISO \(filter.value)"
    case "date":
        return "Date: \(filter.value.replacingOccurrences(of: ",", with: " – "))"
    case "color":
        return "Color: \(filter.value)"
    case "tag":
        return "Tag: \(filter.value)"
    case "near":
        return "Near location"
    case "person":
        return "Person filter"
    case "camera_make":
        return "Camera: \(filter.value)"
    case "camera_model":
        return "Model: \(filter.value)"
    case "lens":
        return "Lens: \(filter.value)"
    case "path":
        return "Path: \(filter.value)"
    default:
        return "\(filter.type): \(filter.value)"
    }
}
